[FIRRTL] Lower RefType Operations to XMR

prithayan opened the pull request:
The `LowerXMRPass`  replaces every `RefResolveOp` with a verbatim expr op encoding the corresponding XMR.
 The pass also removes every value of `RefType` and its uses from the circuit.

This is a dataflow analysis over a very constrained `RefType`. Domain of the dataflow analysis is the set of all `RefSendOp`.
Essentially every `RefType` value must be mapped to one and only one `RefSendOp`.
The analysis propagates the reachable `RefSendOp` to every value of `RefType` across modules.
The `RefResolveOp` is the final leaf into which the dataflow must flow into.

 Due to the downward only reference constraint on XMRs, the post order traversal ensures that the `RefSendOp` will be encountered before any user of the `RefType` value. The pass fails and errors out if every "live" `RefType` value is not mapped to a unique `RefSendOp`.

seldridge reviewed (commented):
I think this algorithm will have issues for multiply instantiated modules with a refsend. It's currently using the refsend as a key when this could be shared if the module with the refsend is multiply instantiated.

The key, I think, needs to be the ref resolve or the algorithm needs to start from the ref resolve and work down towards the send.

WDYT?

prithayan commented:
> I think this algorithm will have issues for multiply instantiated modules with a refsend. It's currently using the refsend as a key when this could be shared if the module with the refsend is multiply instantiated.
> 
> The key, I think, needs to be the ref resolve or the algorithm needs to start from the ref resolve and work down towards the send.
> 
> WDYT?

Thanks @seldridge for the review. For handling multiply instantiated, I think we have to do it in two steps.
First, a post order traversal, to determine the reaching RefSends at any RefResolve.
Now, if there are multiple paths from the RefSend to the RefResolve, we can do a BFS to get the shortest path from the RefResolve to the RefSend.

seldridge commented:
Thinking a bit more... The downwards only constraint should avoid the main thing I was worried about. I.e., all ref resolves for all instances of a module will resolve to the same ref send. Having to do a BFS to determine ownership should be unnecessary.

All I think this algorithm needs is to visit in post order of the instance graph (as it already does), but only visit each module once. (I think it's visiting each module for every instance right now?)

prithayan commented:
> Thinking a bit more... The downwards only constraint should avoid the main thing I was worried about. I.e., all ref resolves for all instances of a module will resolve to the same ref send. Having to do a BFS to determine ownership should be unnecessary.
> 
> All I think this algorithm needs is to visit in post order of the instance graph (as it already does), but only visit each module once. (I think it's visiting each module for every instance right now?)

@seldridge, I updated the algorithm to handle multiple instances. The problem is that if there are multiple paths, we have to trace the unique path from `RefResolve` to `RefSend`. So, we need to store the path to the `RefSend` at every `RefType` value.
Added a testcase to handle this case.

seldridge commented:
Given the downwards reference guarantee, a ref resolve has to match a single ref send (in the module graph) though not necessarily in the instance graph.

I.e., all ref resolves must be the same or we'd have to duplicate modules (which is impossible given the downwards ref guarantee). It may be simpler to only work in module graph space and that would simplify the amount of state that has to be tracked.

This appears to be exactly what the algorithm is doing now by only visiting modules once. 👍

seldridge commented on lib/Dialect/FIRRTL/Transforms/LowerXMR.cpp:
```diff
@@ -0,0 +1,243 @@
+      }
+  }
+
+  ArrayAttr getRemoteRefSend(Value val, Operation *op) {
+    auto iter = reachingRefSendAt.find(val);
+    if (iter != reachingRefSendAt.end())
+      return iter->getSecond();
+    // The referenced module must have already been analyzed, error out if the
+    // dataflow at the child module is not resolved.
+    op->emitOpError(
+        "reference dataflow cannot be traced back to the remote read op");
+    return {};
+  }
```
It would be nice if there was a way to avoid having to pass the `Operation *` as this is only really used in the failing path. It's also annoying to have to do the special casing here if the `val` is a block arg or a has a defining op. I'd consider doing it the latter way.

prithayan replied:
Thanks for the suggestion, it was annoying, updated it to use the special casing, seems to be better.

seldridge commented on lib/Dialect/FIRRTL/Transforms/LowerXMR.cpp:
```diff
@@ -0,0 +1,243 @@
+      auto instanceResult = inst.getResult(portNum);
+      if (!instanceResult.getType().isa<RefType>())
+        continue;
+      if (!refMod) {
+        return inst.emitOpError("cannot lower ext modules with RefType ports");
+      }
```
Nit:
```suggestion
      if (!refMod) 
        return inst.emitOpError("cannot lower ext modules with RefType ports");
```

Great to add this check. We should, going forward, think about how to handle circuits that do this.

seldridge commented on lib/Dialect/FIRRTL/Transforms/LowerXMR.cpp:
```diff
@@ -0,0 +1,243 @@
+                           [&](FModuleOp mod) -> ModuleNamespace & {
+                             return getModuleNamespace(mod);
+                           });
+  }
+  /// Cached module namespaces.
+  DenseMap<Operation *, ModuleNamespace> moduleNamespaces;
```
nit:
```suggestion
  }
  
  /// Cached module namespaces.
  DenseMap<Operation *, ModuleNamespace> moduleNamespaces;
```

seldridge commented on lib/Dialect/FIRRTL/Transforms/LowerXMR.cpp:
```diff
@@ -0,0 +1,243 @@
+  /// of the path exist.
+  DenseMap<Value, ArrayAttr> reachingRefSendAt;
+
+  SmallVector<Operation *> opsToRemove;
```
Nit:
```suggestion
  /// RefResolve, RefSend, and Connects involving them that will be removed.
  SmallVector<Operation *> opsToRemove;
```

seldridge commented on test/Dialect/FIRRTL/lowerXMR.mlir:
```diff
@@ -0,0 +1,67 @@
+
+// -----
+
+// Test for multiple readers and multiple instances
```
Nit:
For maintainability, it may be better to test these features separately from each other.

seldridge commented on lib/Dialect/FIRRTL/Transforms/LowerXMR.cpp:
```diff
@@ -0,0 +1,243 @@
+// instance paths that exist in the circuit.
+// Algorithm:
+//
+// For every RefSendOp
+//  1. The BaseType input is the value to which the final XMR should refer to.
+//  2. Set the `reachingRefSendAt` for the result RefType to an InnerRef to this
+//  op. This map tracks the dataflow path from the original RefSendOp to the
+//  corresponding ref values.
+// For every InstanceOp
+//  1. For every RefType port of the InstanceOp, get the remote RefSendOp
+//     that flows into the corresponding port of the Referenced module.
+//     Because of the order of traversal and the constraints on the ref
+//     ports, the Referenced module ref ports must already be resolved.
+//  2. Update the `reachingRefSendAt` for the corresponding RefSendOp,
+//    to append an InnerRef to this InstanceOp. This denotes that the final
+//    XMR must include this InstanceOp.
+// For every ConnectLike op
+//  1. Copy the dataflow of the src to the dest.
+// For every RefResolveOp,
+//  1. Replace the op result with a VerbatimExpr, representing the XMR.
+//       The InnerRef sequence of symbols from `reachingRefSendAt` is
+//       used to construct the symbol list for the verbatim.
```
Nit: use consistent formatting for items.

seldridge reviewed (approved):
LGTM

prithayan commented:
Thanks @seldridge for the detailed review, addressed all the concerns and Merged the PR.
